File: backend/app/api/v1/projects.py
"""
Project API Router.

Provides REST API endpoints for:
- Project CRUD operations
- Project search and lookup
"""
from typing import Optional, List
from datetime import datetime
from fastapi import APIRouter, Depends, HTTPException, status, Query, Path
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.services.project_service import (
    ProjectService,
    ProjectServiceError,
    ProjectNotFoundError,
    ProjectDuplicateCodeError
)
from app.schemas.project import (
    ProjectCreate, ProjectUpdate,
    ProjectSearchParams,
    ProjectResponse, ProjectDetailResponse,
    ProjectListResponse, ProjectSummary
)

router = APIRouter(prefix="/projects", tags=["Projects"])


# ==========================================================================
# Dependency Injection
# ==========================================================================

async def get_project_service(db: AsyncSession = Depends(get_db)) -> ProjectService:
    """Get project service instance."""
    return ProjectService(db)


# ==========================================================================
# Exception Handler Helper
# ==========================================================================

def handle_project_error(error: ProjectServiceError) -> HTTPException:
    """Convert ProjectServiceError to appropriate HTTPException."""
    status_code = status.HTTP_400_BAD_REQUEST
    error_code = "PROJECT_ERROR"

    if isinstance(error, ProjectNotFoundError):
        status_code = status.HTTP_404_NOT_FOUND
        error_code = "PROJECT_NOT_FOUND"
    elif isinstance(error, ProjectDuplicateCodeError):
        status_code = status.HTTP_409_CONFLICT
        error_code = "PROJECT_DUPLICATE_CODE"

    return HTTPException(
        status_code=status_code,
        detail={
            "success": False,
            "error": {
                "code": error_code,
                "message": str(error)
            }
        }
    )


# ==========================================================================
# Project CRUD Endpoints
# ==========================================================================

@router.post(
    "",
    response_model=ProjectDetailResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create a new project",
    description="""
    Create a new project.

    A project requires:
    - Code (unique within society)
    - Name
    - Client ID
    - Payment condition ID
    - Payment mode ID
    - VAT rate ID
    - Society ID
    """
)
async def create_project(
    data: ProjectCreate,
    service: ProjectService = Depends(get_project_service)
):
    """Create a new project."""
    try:
        project = await service.create_project(data)
        return project
    except ProjectServiceError as e:
        raise handle_project_error(e)


@router.get(
    "",
    response_model=ProjectListResponse,
    summary="Search and list projects",
    description="""
    Search projects with optional filters and pagination.

    Supports filtering by:
    - Search term (matches code and name)
    - Code (partial match)
    - Name (partial match)
    - Client ID
    - Society ID
    - Creator ID
    - Date range
    """
)
async def search_projects(
    search: Optional[str] = Query(None, description="Search term"),
    code: Optional[str] = Query(None, description="Filter by code (partial match)"),
    name: Optional[str] = Query(None, description="Filter by name (partial match)"),
    client_id: Optional[int] = Query(None, description="Filter by client ID"),
    society_id: Optional[int] = Query(None, description="Filter by society ID"),
    creator_id: Optional[int] = Query(None, description="Filter by creator ID"),
    date_from: Optional[datetime] = Query(None, description="Created from date"),
    date_to: Optional[datetime] = Query(None, description="Created to date"),
    page: int = Query(1, ge=1, description="Page number"),
    page_size: int = Query(20, ge=1, le=100, description="Items per page"),
    sort_by: str = Query("prj_d_creation", description="Sort field"),
    sort_order: str = Query("desc", description="Sort order (asc/desc)"),
    service: ProjectService = Depends(get_project_service)
):
    """Search and list projects with pagination."""
    params = ProjectSearchParams(
        search=search,
        code=code,
        name=name,
        client_id=client_id,
        society_id=society_id,
        creator_id=creator_id,
        date_from=date_from,
        date_to=date_to,
        page=page,
        page_size=page_size,
        sort_by=sort_by,
        sort_order=sort_order
    )
    return await service.search_projects(params)


@router.get(
    "/lookup",
    response_model=List[dict],
    summary="Get projects for lookup/dropdown",
    description="Get lightweight project data for selection in dropdowns."
)
async def get_project_lookup(
    soc_id: int = Query(..., description="Society ID"),
    client_id: Optional[int] = Query(None, description="Filter by client ID"),
    search: Optional[str] = Query(None, description="Search term"),
    limit: int = Query(50, ge=1, le=100, description="Maximum records"),
    service: ProjectService = Depends(get_project_service)
):
    """Get projects for dropdown/lookup."""
    return await service.get_project_lookup(soc_id, client_id, search, limit)


@router.get(
    "/summary",
    response_model=ProjectSummary,
    summary="Get project summary statistics",
    description="Get summary statistics for projects."
)
async def get_project_summary(
    soc_id: Optional[int] = Query(None, description="Filter by society ID"),
    service: ProjectService = Depends(get_project_service)
):
    """Get project summary statistics."""
    return await service.get_project_summary(soc_id)


@router.get(
    "/by-client/{client_id}",
    response_model=List[ProjectResponse],
    summary="Get projects by client",
    description="Get all projects for a specific client."
)
async def get_projects_by_client(
    client_id: int = Path(..., description="Client ID"),
    skip: int = Query(0, ge=0, description="Number of records to skip"),
    limit: int = Query(50, ge=1, le=100, description="Maximum records"),
    service: ProjectService = Depends(get_project_service)
):
    """Get all projects for a client."""
    return await service.get_projects_by_client(client_id, skip, limit)


@router.get(
    "/by-society/{soc_id}",
    response_model=List[ProjectResponse],
    summary="Get projects by society",
    description="Get all projects for a specific society."
)
async def get_projects_by_society(
    soc_id: int = Path(..., description="Society ID"),
    skip: int = Query(0, ge=0, description="Number of records to skip"),
    limit: int = Query(50, ge=1, le=100, description="Maximum records"),
    service: ProjectService = Depends(get_project_service)
):
    """Get all projects for a society."""
    return await service.get_projects_by_society(soc_id, skip, limit)


@router.get(
    "/{project_id}",
    response_model=ProjectDetailResponse,
    summary="Get project by ID",
    description="Get detailed information about a specific project with resolved lookup names."
)
async def get_project(
    project_id: int = Path(..., description="Project ID"),
    service: ProjectService = Depends(get_project_service)
):
    """Get a project by ID with resolved lookup names."""
    try:
        return await service.get_project_detail(project_id)
    except ProjectServiceError as e:
        raise handle_project_error(e)


@router.put(
    "/{project_id}",
    response_model=ProjectDetailResponse,
    summary="Update a project",
    description="Update an existing project."
)
async def update_project(
    project_id: int = Path(..., description="Project ID"),
    data: ProjectUpdate = ...,
    service: ProjectService = Depends(get_project_service)
):
    """Update a project."""
    try:
        return await service.update_project(project_id, data)
    except ProjectServiceError as e:
        raise handle_project_error(e)


@router.delete(
    "/{project_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete a project",
    description="Delete a project by ID."
)
async def delete_project(
    project_id: int = Path(..., description="Project ID"),
    service: ProjectService = Depends(get_project_service)
):
    """Delete a project."""
    try:
        await service.delete_project(project_id)
    except ProjectServiceError as e:
        raise handle_project_error(e)

File: frontend/src/api/projects.ts
import apiClient from './client'
import type {
  Project,
  ProjectCreateDto,
  ProjectUpdateDto,
  ProjectSearchParams,
  ProjectLookupItem,
  ProjectSummary,
  ProjectDetailResponse,
  mapProjectResponseToProject,
} from '@/types/project'
import type { PagedResponse } from '@/types/api'

/**
 * Transform backend response to frontend format
 */
function transformProject(data: ProjectDetailResponse): Project {
  return {
    id: data.prj_id,
    code: data.prj_code,
    name: data.prj_name,
    displayName: data.display_name || `${data.prj_code} - ${data.prj_name}`,
    clientId: data.cli_id,
    clientName: data.client_name,
    societyId: data.soc_id,
    societyName: data.society_name,
    creatorId: data.usr_creator_id,
    creatorName: data.creator_name,
    paymentTermId: data.pco_id,
    paymentTermName: data.payment_term_name,
    paymentModeId: data.pmo_id,
    paymentModeName: data.payment_mode_name,
    vatRateId: data.vat_id,
    vatRateName: data.vat_rate_name,
    createdAt: data.prj_d_creation,
    updatedAt: data.prj_d_update,
  }
}

/**
 * Projects API methods
 */
export const projectsApi = {
  /**
   * Get paginated list of projects with optional filtering
   */
  async getAll(params: ProjectSearchParams = {}): Promise<PagedResponse<Project>> {
    const queryParams: Record<string, any> = {}
    
    if (params.search) queryParams.search = params.search
    if (params.code) queryParams.code = params.code
    if (params.name) queryParams.name = params.name
    if (params.client_id) queryParams.client_id = params.client_id
    if (params.society_id) queryParams.society_id = params.society_id
    if (params.creator_id) queryParams.creator_id = params.creator_id
    if (params.date_from) queryParams.date_from = params.date_from
    if (params.date_to) queryParams.date_to = params.date_to
    if (params.page) queryParams.page = params.page
    if (params.page_size) queryParams.page_size = params.page_size
    if (params.sort_by) queryParams.sort_by = params.sort_by
    if (params.sort_order) queryParams.sort_order = params.sort_order

    const response = await apiClient.get('/projects', { params: queryParams })
    const data = response.data
    const page = data.page || params.page || 1
    const pageSize = data.page_size || params.page_size || 20
    const totalCount = data.total || 0
    const totalPages = data.total_pages || Math.ceil(totalCount / pageSize)

    return {
      success: true,
      data: (data.items || []).map(transformProject),
      page,
      pageSize,
      totalCount,
      totalPages,
      hasNextPage: page < totalPages,
      hasPreviousPage: page > 1,
    }
  },

  /**
   * Get a single project by ID
   */
  async getById(id: number): Promise<Project> {
    const response = await apiClient.get(`/projects/${id}`)
    return transformProject(response.data)
  },

  /**
   * Create a new project
   */
  async create(data: ProjectCreateDto): Promise<Project> {
    const response = await apiClient.post('/projects', data)
    return transformProject(response.data)
  },

  /**
   * Update an existing project
   */
  async update(id: number, data: ProjectUpdateDto): Promise<Project> {
    const response = await apiClient.put(`/projects/${id}`, data)
    return transformProject(response.data)
  },

  /**
   * Delete a project
   */
  async delete(id: number): Promise<void> {
    await apiClient.delete(`/projects/${id}`)
  },

  /**
   * Get projects by client ID
   */
  async getByClient(clientId: number, skip = 0, limit = 50): Promise<Project[]> {
    const response = await apiClient.get(`/projects/by-client/${clientId}`, {
      params: { skip, limit },
    })
    return (response.data || []).map(transformProject)
  },

  /**
   * Get projects for dropdown lookup
   */
  async getLookup(
    societyId: number,
    clientId?: number,
    search?: string,
    limit = 50
  ): Promise<ProjectLookupItem[]> {
    const params: Record<string, any> = { soc_id: societyId, limit }
    if (clientId) params.client_id = clientId
    if (search) params.search = search

    const response = await apiClient.get('/projects/lookup', { params })
    return (response.data || []).map((item: any) => ({
      id: item.prj_id || item.id,
      code: item.prj_code || item.code,
      name: item.prj_name || item.name,
      display_name: item.display_name || `${item.prj_code || item.code} - ${item.prj_name || item.name}`,
      client_id: item.cli_id || item.client_id,
    }))
  },

  /**
   * Get project summary statistics
   */
  async getSummary(societyId?: number): Promise<ProjectSummary> {
    const params: Record<string, any> = {}
    if (societyId) params.soc_id = societyId

    const response = await apiClient.get('/projects/summary', { params })
    return {
      total_count: response.data.total_count || 0,
      projects_by_client: response.data.projects_by_client || 0,
      recent_projects: response.data.recent_projects || 0,
    }
  },
}
